File: WordBreakII.py
class Solution:
    # @param s, a string
    # @param dict, a set of string
    # @return a list of strings
    def wordBreak(self, s, dict):
        if len(s) == 0:
            if "" in dict:
                return [""]
            else:
                return []
        booleans = [False for i in range(len(s))]
        sentences = [set() for i in range(len(s))]
        if s[0] in dict:
            booleans[0] = True
            sentences[0].add(s[0])
        for pos in range(1, len(s)):
            if s[:pos+1] in dict:
                booleans[pos] = True
                sentences[pos].add(s[:pos+1])
            for sub in range(0, pos):
                if booleans[sub] and s[sub+1:pos+1] in dict:
                    booleans[pos] = True
                    for substr in sentences[sub]:
                        if substr != "":
                            sentences[pos].add(substr + " " + s[sub+1:pos+1])
                        else:
                            sentences[pos].add(s[sub+1 : pos+1])
        return list(sentences[len(s)-1])

File: WordBreakII.cc
#include <iostream>
#include <string>
#include <vector>
#include <unordered_set>

using namespace std;

class Solution {
    public:
        vector<string> wordBreak(string s, unordered_set<string> &dict) {
            vector<vector<string>> broken(s.size(), vector<string>());
            for (int i = s.size() - 1; i>= 0; i--) {
                string current = s.substr(i, s.size() - i);
                if (dict.find(current) != dict.end()) {
                    broken[i].push_back(current);
                }

                for (int j = i + 1; j < s.size(); j++) {
                    string first = s.substr(i, j - i);
                    string second = s.substr(j, s.size() - j);
                    if (!broken[j].empty() && dict.find(first) != dict.end()) {
                        for (auto sub : broken[j]) {
                            if (sub == "") {
                                broken[i].push_back(first);
                            } else {
                                broken[i].push_back(first + " " + sub);
                            }
                        }
                    }
                }
            }

            return broken[0];
        }
};

int main()
{
    Solution sol;
    unordered_set<string> dict = {"a", "b"};
    vector<string> res = sol.wordBreak("ab", dict);
}
